tools/release-ci/cmd/sidecar.go

```diff
@@ -0,0 +1,88 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/release/tools/release-ci/pkg/logging/gcs"
+)
+
+// sidecarCmd should run alongside a test pod and uploads files to GCS
+var sidecarCmd = &cobra.Command{
+	Use:   "sidecar",
+	Short: "Uploads build log and artifacts to GCS",
+	Long: `Uploads build log and artifacts to GCS
+
+The GCS sidecar container will run alongside the container that
+wraps its process with the entrypoint. These two processes are
+configured together; the sidecar will look for the output from
+this entrypoint to upload to GCS.`,
+	Args: cobra.NoArgs(),
+	Run: func(cmd *cobra.Command, args []string) {
+		if err := runSidecar(args); err != nil {
+			fmt.Printf("Error: %v\n", err)
+			os.Exit(1)
+		}
+	},
+}
+
+func init() {
+	RootCmd.AddCommand(sidecarCmd)
+	sidecarCmd.Flags().StringVar(&configurationFile, "config", "", "The location of the configuration file")
+}
+
+func runSidecar(_ []string) error {
+	config, err := loadConfig(configurationFile)
+	if err != nil {
+		return err
+	}
+
+	gcsBucket, err := createGcsClient(config.GcsBucket)
+	if err != nil {
+		return err
+	}
+
+	if err := gcs.UploadStartingData(config.ConfigurationFile, gcsBucket); err != nil {
+		return err
+	}
+
+	returnCode, err := retrieveProcessReturnCode(config.MarkerFile)
+	if err != nil {
+		// even if we fail to determine this info,
+		// we want to upload as much as we can
+		fmt.Printf("failed to determine process return code: %v", err)
+	}
+	passed := returnCode == 0 && err == nil
+
+	return gcs.UploadFinishedData(config.ProcessLog, config.ConfigurationFile, config.ArtifactDir, passed, gcsBucket)
+}
+
+func createGcsClient(bucket string) (*storage.BucketHandle, error) {
+	ctx := context.Background()
+	gcsClient, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to GCS: %v", err)
+	}
+	return gcsClient.Bucket(bucket), nil
+}
+
+// retrieveProcessReturnCode waits until the return code
+// is available in the marker file, then returns it
+func retrieveProcessReturnCode(markerFile string) (int, error) {
```

Comment:
Refactoring this to use `fsnotify` is the last bit -- other than that, this could use some eyes @kargakis @csrwng